hack/verify-canonical-import-path.sh

```diff
@@ -0,0 +1,50 @@
+#!/usr/bin/env bash
+
+# Copyright 2018 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
+staging_repos="${KUBE_ROOT}/staging/src/k8s.io"
+
+# verify-canonical-import-path checks that directories in staging
+# having doc.go contain canonical import paths.
+# See https://golang.org/doc/go1.4#canonicalimports for more details.
+
+failed=false
+while read i; do
+    # don't consider generated files
+    if grep -q "DO NOT EDIT" $i; then
+        continue
+    fi
+
+    # extract import path and package name from the file name
+    # If i="hack/../staging/src/k8s.io/client-go/scale/doc.go", then
+    # import_path=k8s.io/client-go/scale and package_name=scale
+    import_path=$(echo -n ${i:20} | head -c-7)
```

Comment:
Earlier, this line was `import_path=${i:20:-7}`. Looks like negative substring lengths were added in Bash 4:

- http://tldp.org/LDP/abs/html/string-manipulation.html#SUBSTREXTR01
- http://tldp.org/LDP/abs/html/bashver4.html

@noqcks I've updated the line to use `head` now. Can you verify if this works on Bash 3? I'm on Bash 4.4.

Reply:
getting `head: illegal byte count -- -7`

Reply:
@noqcks Updated. Can you check again? This should work. :grimacing:

Reply:
Looking 👍  now

However, im running `./hack/verify-canonical-import-path.sh` from project root and getting some false positives of `does not use canonical import paths.` I believe... 

On my branch https://github.com/kubernetes/kubernetes/pull/68266#issuecomment-431972596 the following show for example. 
```
staging/src/c/k8s.io/client-go/tools/metrics/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/bootstrap/token/util/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/record/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/cache/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/clientcmd/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/clientcmd/api/v1/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/clientcmd/api/latest/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/clientcmd/api/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/leaderelection/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/leaderelection/resourcelock/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/watch/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/reference/doc.go does not use canonical import paths.
staging/src/c/k8s.io/client-go/tools/pager/doc.go does not use canonical import paths.
```

But maybe I am using this incorrectly..